tests/helpers/fixed_string.h

```diff
@@ -0,0 +1,290 @@
+  /// the string, not null terminated
+  CharType storage[N];
+
+  using type = CharType;
+  static constexpr std::endian endianess = endianess_;
+
+  constexpr CTString() = default;
+
+  /// construction from a string literal (includes a null character)
+  constexpr explicit CTString(CharType const (&str)[N + 1]) {
+    std::copy(str, str + size(), storage);
+
+    if constexpr (std::is_same_v<CharType, char>) {
+      for (const unsigned char c : storage) {
+        if (c > 0x7F) {
+          throw "Not latin1!";
+        }
+      }
+    }
+  }
+
+  /// construct from a range with the appropriate type
+  constexpr explicit CTString(std::ranges::input_range auto &&range)
+    requires detail::valid_chartype<
+                 std::ranges::range_value_t<decltype(range)>> &&
+             std::ranges::sized_range<decltype(range)>
+  {
+  #if !defined(_MSVC_LANG) && !defined(__clang__)
+    constexpr auto Ninput = std::ranges::size(range);
+    static_assert(Ninput == size());
+  #endif
+    std::ranges::copy(range, storage);
+    if constexpr (std::is_same_v<CharType, char>) {
+      for (const unsigned char c : storage) {
+        if (c > 0x7F) {
+          throw "Not latin1!";
+        }
+      }
+    }
+  }
+
+  constexpr bool operator==(const CTString &other) const noexcept = default;
+
+  /// size in number of elements
+  static constexpr std::size_t size() noexcept { return N; }
+  static constexpr bool empty() noexcept { return size() == 0; }
+  constexpr CharType *data() noexcept { return storage; }
+  constexpr const CharType *data() const noexcept { return storage; }
+  constexpr auto span() noexcept { return std::span(storage); }
+  constexpr auto span() const noexcept { return std::span(storage); }
+
+  constexpr auto begin() noexcept { return data(); }
+  constexpr auto end() noexcept { return data() + size(); }
+  constexpr auto begin() const noexcept { return data(); }
+  constexpr auto end() const noexcept { return data() + size(); }
+  constexpr CharType operator[](auto index) const noexcept {
+    return storage[index];
+  }
+  constexpr CharType &operator[](auto index) noexcept { return storage[index]; }
+  template <typename DestCharType> constexpr auto as_array() const noexcept {
+    std::array<DestCharType, size()> ret;
+    for (std::size_t i = 0; i < size(); ++i) {
+      ret[i] = static_cast<DestCharType>(storage[i]);
+    }
+    return ret;
+  }
+};
+
+/// Class template argument deduction (CTAD) for a string literal (null
+/// terminated)
+template <detail::valid_chartype CharType, std::size_t N>
+CTString(const CharType (&str)[N]) -> CTString<CharType, N - 1>;
+
+/// Class template argument deduction (CTAD) for a range (must have size known
+/// at compile time)
+template <std::ranges::input_range R>
+CTString(R &&)
+    -> CTString<std::ranges::range_value_t<R>, std::ranges::size(R{})>;
+
+/// creates a fixed size string of UTF-32
+template <detail::TmpStringLiteral a> constexpr auto operator""_utf32() {
```

Comment:
This is all fine, but one idea to consider.

Here is something I have always wanted...

```cpp
constexpr char * data = "AaAa"_base64; 
constexpr uint8_t * data = "AaAa"_base64; 
```

So that you can just past (hardcoded) some base64 content in your C++ code and it gets decoded (at compile time) to the corresponding binary data.

(This not what you are working towards here, but I want to bring it up.)